libflux/core/src/semantic/nodes.rs

```diff
@@ -1013,278 +1460,359 @@ impl BinaryExpr {
         // Do this first so that we can return an error if one occurs.
         let (env, lcons) = self.left.infer(env, f)?;
         let (env, rcons) = self.right.infer(env, f)?;
-
+        
+        let u = MonoType::Var(f.fresh());
+        let v = MonoType::Var(f.fresh());
         let cons = match self.operator {
             // The following operators require both sides to be equal.
             ast::Operator::AdditionOperator => Constraints::from(vec![
-                Constraint::Equal {
-                    exp: self.left.type_of(),
-                    act: self.right.type_of(),
-                    loc: self.right.loc().clone(),
-                },
-                Constraint::Equal {
+            	Constraint::Equal {
```

Comment:
We already talked about this, but I just wanted to document it. We either need another equality constraint tying the operand types to the type of the binary expression itself, or alternatively we need to update `Expression::type_of()` to return one of its operand types instead of the `typ` field.

I'm actually leaning towards the latter. The reason being that we could then remove the `typ` field from many of the nodes. This will significantly reduce the size of the semantic graph since instead of storing the type info with each node, we'll be computing it on the fly.